### ISSUE TITLE
[Bug]: Stale partial prefix-cache hash resurrected into the cache after full-block promotion

### ISSUE THREAD

**Issue by ErenAta16** [vllm-project/vllm#49125]
### Your current environment

This is a pure Python bug in `BlockPool.cache_partial_block()` / `FullAttentionManager._cache_partial_tail_block()` (`vllm/v1/core/block_pool.py`, `vllm/v1/core/single_type_kv_cache_manager.py`). It has no hardware/kernel/CUDA dependency, so `collect_env.py` isn't meaningful here. Verified against `vllm-project/vllm` main at commit `ace9fda495bc9a132584e054cc909d504164141f` (2026-07-19), using the real `BlockPool`, `FullAttentionManager`, `Request`, and `FullAttentionSpec` classes, no mocking of the KV-cache logic under test. The reproduction runs standalone with no GPU; the only stubbing is for unrelated platform-detection import plumbing on a non-editable-install checkout (noted inline in the script).

This affects the fine-grained/partial prefix-caching feature added by #45939 and #46384 (merged 2026-06-22 and 2026-07-12), which is very recent.

### 🐛 Describe the bug

**Summary:** when a request's KV-cache block already holds a partial-prefix-cache hash entry (registered by `FullAttentionManager._cache_partial_tail_block()` for the token boundary where the prompt ends inside a block), and normal decoding later fills that same block up completely so it gets promoted to a full-block hash, the very next call to `cache_blocks()` — which happens on every subsequent scheduling step for the lifetime of the request — re-registers the old, already-superseded partial hash back into the prefix cache. This directly contradicts the invariant the project's own test suite asserts immediately after promotion.

**Where**

`vllm/v1/core/single_type_kv_cache_manager.py`, `FullAttentionManager`:

```python
    def cache_blocks(
        self,
        request: Request,
        num_tokens: int,
        retention_interval: int | None = None,
    ) -> None:
        super().cache_blocks(request, num_tokens, retention_interval=retention_interval)
        hash_block_size = self.block_pool.hash_block_size
        if self.block_size == hash_block_size:
            return
        self._cache_partial_tail_block(request, num_tokens)

    def _cache_partial_tail_block(
        self,
        request: Request,
        num_tokens: int,
    ) -> None:
        """Cache the prompt tail when it ends inside a cache block.
        ...
        """
        hash_block_size = self.block_pool.hash_block_size
        boundary_tokens = request.num_prompt_tokens // hash_block_size * hash_block_size
        if boundary_tokens == 0 or boundary_tokens > num_tokens:
            return
        if boundary_tokens % self.block_size == 0:
            return

        blocks = self.req_to_blocks[request.request_id]
        block_idx = boundary_tokens // self.block_size
        if block_idx >= len(blocks):
            return
        self.block_pool.cache_partial_block(
            request=request,
            block=blocks[block_idx],
            num_tokens=boundary_tokens,
            kv_cache_group_id=self.kv_cache_group_id,
            block_size=self.block_size,
        )
```

(current lines 753–793)

`boundary_tokens` is derived only from `request.num_prompt_tokens`, which is fixed for the life of the request. Nothing in `_cache_partial_tail_block` checks whether the target block (`blocks[block_idx]`) has, in the meantime, been promoted to a full-block hash by `cache_full_blocks()` (called just above it, inside `super().cache_blocks()`, in the very same call). Since `cache_blocks()` is invoked on every `allocate_slots()` call — i.e. every scheduling step, including ordinary decode steps (`vllm/v1/core/kv_cache_manager.py:502`) — `_cache_partial_tail_block` keeps re-running with the same stale `boundary_tokens` value for as long as the request exists.

`vllm/v1/core/block_pool.py`, `BlockPool.cache_partial_block()` (current lines 445–544) doesn't guard against this either:

```python
        already_cached = block.block_hash == block_hash_with_group_id or (
            self.cached_block_hash_to_block.contain(
                block_hash_with_group_id, block.block_id
            )
        )
        if (
            not already_cached
            and block.block_hash is not None
            and block.block_hash_num_tokens is not None
            and block.block_hash_num_tokens < num_hash_blocks * self.hash_block_size
        ):
            removed_hashes = self._remove_cached_block_hashes(block)
            self._emit_block_removed_events(removed_hashes)
        self._insert_block_hash(
            block_hash_with_group_id,
            block,
            num_tokens=num_hash_blocks * self.hash_block_size,
        )
```

Once the block has been promoted, `block.block_hash` is the new, larger full-block hash and `block.block_hash_num_tokens` is its (larger) token count. The stale call's `already_cached` check is `False` (the old partial hash's cache-map entry was already removed at promotion time), and the removal-guard condition is also `False` (`block.block_hash_num_tokens < num_hash_blocks * self.hash_block_size` — the *existing* hash has *more* tokens than the *stale* one being re-registered, so nothing gets removed). Execution falls through to `_insert_block_hash(...)`, which — since `block.block_hash` no longer equals the stale key and the cache map no longer contains it — re-inserts the stale hash as a *secondary* entry in `cached_block_hash_to_block` and `cached_block_hashes_by_block`, pointing at the same block.

The result: `get_cached_block()` on the old, already-superseded partial-boundary hash starts returning the block again, even though the block's real current state is the (larger) promoted full-block hash. If `enable_kv_cache_events` is on, this also emits a spurious `BlockStored` event for a hash the server already told external consumers (via the promotion's `BlockRemoved` event) it had discarded — external KV-cache-aware routers/gateways that mirror server state from these events end up with a stale, contradictory view of what's cached where.

### Reproduction

Real, deterministic reproduction against the actual, unmodified `BlockPool` and `FullAttentionManager` (no mocking of the manager/pool logic; only unrelated import plumbing is stubbed for a non-GPU, non-editable-install checkout, noted inline). Save as `repro.py` and run with `python repro.py /path/to/vllm`:

```python
import os
import sys
import types

VLLM_ROOT = sys.argv[1] if len(sys.argv) > 1 else "."
os.environ["VLLM_LOGGING_LEVEL"] = "WARNING"
sys.path.insert(0, VLLM_ROOT)

# --- Unrelated import-plumbing stubs (Windows / non-editable-install only) ---
sys.modules["uvloop"] = types.ModuleType("uvloop")
sys.modules["uvloop"].run = lambda *a, **k: None
try:
    import xgrammar  # noqa: F401
except ImportError:
    xg = types.ModuleType("xgrammar")
    xg.GrammarMatcher = type("GrammarMatcher", (), {})
    xg.CompiledGrammar = type("CompiledGrammar", (), {})
    sys.modules["xgrammar"] = xg

import torch
from vllm.sampling_params import SamplingParams
from vllm.utils.hashing import sha256
from vllm.v1.core.block_pool import BlockPool
from vllm.v1.core.kv_cache_utils import (
    BlockHashListWithBlockSize, get_request_block_hasher, init_none_hash,
)
from vllm.v1.core.single_type_kv_cache_manager import FullAttentionManager
from vllm.v1.kv_cache_interface import FullAttentionSpec
from vllm.v1.request import Request

init_none_hash(sha256)


def make_request(request_id, prompt_token_ids, hash_block_size):
    sp = SamplingParams(max_tokens=17)
    sp.update_from_generation_config({}, eos_token_id=100)
    return Request(request_id=request_id, prompt_token_ids=prompt_token_ids,
                    sampling_params=sp, pooling_params=None,
                    block_hasher=get_request_block_hasher(hash_block_size, sha256))


def boundary_hash(req, hash_block_size, num_tokens):
    return req.block_hashes[num_tokens // hash_block_size - 1]


def main():
    hash_block_size = 2
    block_size = 6
    kv_cache_group_id = 0

    prompt_token_ids = [0, 0, 1, 1, 2, 2, 3, 3, 4, 4]  # 10 tokens
    req = make_request("R", prompt_token_ids, hash_block_size)

    pool = BlockPool(num_gpu_blocks=4, enable_caching=True, hash_block_size=hash_block_size)
    kv_cache_spec = FullAttentionSpec(block_size=block_size, num_kv_heads=1,
                                       head_size=1, dtype=torch.float32)
    manager = FullAttentionManager(kv_cache_spec=kv_cache_spec, block_pool=pool,
                                    enable_caching=True, kv_cache_group_id=kv_cache_group_id,
                                    scheduler_block_size=block_size)
    manager.req_to_blocks[req.request_id] = pool.get_new_blocks(2)

    # Prompt (10 tokens) finishes prefill: block 0 full, block 1 partially
    # filled (tokens 6-10). cache_blocks() registers the partial-tail hash.
    manager.cache_blocks(req, num_tokens=10)
    partial_hash_10 = boundary_hash(req, hash_block_size, 10)
    block1 = manager.req_to_blocks[req.request_id][1]
    assert pool.get_cached_block(partial_hash_10, [kv_cache_group_id]) == [block1]

    # Two ordinary decode steps append tokens 10 and 11, filling block 1
    # completely (10 -> 12 tokens). cache_blocks() promotes block 1 to a full
    # 12-token hash IN THE SAME CALL that re-runs _cache_partial_tail_block().
    req.append_output_token_ids([5])
    manager.cache_blocks(req, num_tokens=11)
    req.append_output_token_ids([5])
    manager.cache_blocks(req, num_tokens=12)

    full_hashes = BlockHashListWithBlockSize(req.block_hashes, hash_block_size, block_size)
    promoted_full_hash = full_hashes[1]
    hit_full = pool.get_cached_block(promoted_full_hash, [kv_cache_group_id])
    hit_stale = pool.get_cached_block(partial_hash_10, [kv_cache_group_id])
    print(f"full hash present = {hit_full == [block1]}, stale 10-token hash present = {hit_stale}")

    if hit_stale == [block1]:
        print("BUG CONFIRMED: stale partial hash resurrected after promotion.")
        sys.exit(1)
    sys.exit(0)


if __name__ == "__main__":
    main()
```

**Actual output when run against the real, unmodified `BlockPool`/`FullAttentionManager`:**

```
full hash present = True, stale 10-token hash present = [KVCacheBlock(block_id=2, ref_cnt=1, _block_hash=b'...', _block_hash_num_tokens=12, prev_free_block=None, next_free_block=None)]
BUG CONFIRMED: stale partial hash resurrected after promotion.
```

The stale 10-token partial hash resolves to the same block object even though that block's real, current state is the promoted 12-token full-block hash — the resurrection happens on the very step the block naturally completes, not on some later step.

### Expected behavior

Once a block has been promoted to a full-block hash, `_cache_partial_tail_block()` should not re-register the old partial boundary that the promotion already superseded. `get_cached_block(old_partial_hash, ...)` should keep returning `None` after promotion, matching the invariant the project's own `test_partial_block_promotes_to_direct_full_block_hash` test (`tests/v1/core/prefix_cache/test_partial_prefix_cache_primitives.py`) asserts immediately after promotion — that test just doesn't cover the case where `cache_blocks()`/`_cache_partial_tail_block()` runs again afterward, which happens unconditionally on every subsequent scheduling step in production.

### Actual behavior

The stale, already-superseded partial-boundary hash gets re-inserted into `cached_block_hash_to_block` and `cached_block_hashes_by_block` as a secondary entry on the block, resurrecting a prefix-cache key that the server had already discarded. Concretely this means:
- `cached_block_hashes_by_block` on the promoted block permanently carries an extra stale entry until the block is evicted (it doesn't grow further since later calls become idempotent once the stale key is back in the map, but the code path is silently masking a real correctness gap).
- With `enable_kv_cache_events=True`, a spurious `BlockStored` event is emitted for a hash the server already reported removed via the promotion's `BlockRemoved` event — external KV-cache-aware routers/gateways consuming these events end up with a cache-state view that disagrees with the server's actual state.
- A later request whose token sequence happens to hash-chain-match the stale boundary can get a "hit" through this resurrected key. The engine's copy-on-write path for genuine cross-request partial hits (`_apply_cow` / `move_block_hashes`) does still protect against data corruption in that specific scenario, but the underlying cache-index state is nonetheless wrong and the KV-events contract is broken regardless.

### Suggested fix

In `FullAttentionManager._cache_partial_tail_block()` (`vllm/v1/core/single_type_kv_cache_manager.py`, current line ~765), skip re-registering the partial entry once the target block has already been promoted — e.g. bail out early if `blocks[block_idx].block_hash_num_tokens` is already `>= boundary_tokens` before calling `self.block_pool.cache_partial_block(...)`, similar to the "already cached" fast-paths used elsewhere in this file. Alternatively, `BlockPool.cache_partial_block()` (`vllm/v1/core/block_pool.py`, current line 445) could reject/no-op when `block.block_hash_num_tokens` (the block's current, larger hash) already covers more tokens than the incoming (smaller/stale) `num_tokens`, instead of only guarding the "block hash is smaller than the new one" growth case.

### Are you a ML Ops Team?

No

### What vLLM version are you on?

`main` branch, verified against commit `ace9fda495bc9a132584e054cc909d504164141f` (2026-07-19). The fine-grained/partial prefix-caching feature this affects was introduced by #45939 (merged 2026-06-22) and extended by #46384 (merged 2026-07-12), so this is present in any build containing those changes.

**Comment by neetp7k9** [comment]
I've reproduced this bug on latest main (583a00257d, vLLM 0.27.2rc1) and have a fix ready. Working on a PR.

**Comment by ErenAta16** [comment]
Good to hear it reproduces on `583a0025` as well, that is a month of drift since the commit I filed against and the path has been changing quickly. The fix is yours, I am not working on one.

Three things that might save you time, since I spent a while in this code.

**There are two places the guard can go, and they are not equivalent.** In `_cache_partial_tail_block` you can compare the target block's current state against `boundary_tokens` and return early. In `cache_partial_block` you can refuse to insert a hash that is smaller than the one the block already carries. The second is the stronger invariant, because it holds for any future caller rather than just this one, but it is also the one that has to be careful not to reject the legitimate first registration, where `block.block_hash` is either `None` or the same partial hash. The existing `already_cached` check covers the re-entry case but not the promoted case, which is why the stale call falls through.

**The removal guard is the specific line that lets it through.** `block.block_hash_num_tokens < num_hash_blocks * self.hash_block_size` is comparing the wrong direction for this scenario: after promotion the block's existing hash covers *more* tokens than the stale one being re-registered, so the condition is false and nothing gets removed before `_insert_block_hash` runs. Whatever shape the fix takes, that asymmetry is worth a comment, because reading it the condition looks like it is already handling exactly this case.

**Do not forget the event half.** The re-insert also emits a `BlockStored` for a hash the server already announced as removed at promotion time. A fix that stops `get_cached_block` returning the stale block but still emits the event would leave external KV-aware routers with the same contradictory view, which is the part that actually reaches users. Worth an assertion on the emitted events, not only on the cache map.

On testing: the repro in the issue runs standalone with no GPU and drives the real `BlockPool` and `FullAttentionManager` rather than mocks, so it should adapt into a regression test without needing hardware in CI. Take any of it you find useful, no attribution needed. If it helps, I am happy to run your branch against it and report back before you send the PR.

**Comment by neetp7k9** [comment]
@ErenAta16 thanks — unusually useful review, all three points landed. The PR is already up at #52972 (opened 08-19, so you likely hadn't seen it). Just pushed a second commit addressing your feedback.

**The event half — you were right, and my test covered only half the bug.** I'd asserted on `get_cached_block` and stopped. With the guard removed the emitted sequence is:

```
BlockRemoved(partial) -> BlockStored(full) -> BlockStored(partial)
```

so the retired hash is re-announced as stored on the same `cache_blocks` call that removed it. The fix does suppress the event (it early-returns before `cache_partial_block`, so nothing is queued), but nothing was pinning that. The regression test now runs the pool with `enable_kv_cache_events=True` and asserts (a) no `BlockStored` re-announces the partial hash after its `BlockRemoved`, and (b) steady-state decode inside an already-promoted block emits no events at all.

**The removal guard.** Agreed the direction is confusing. `block_hash_num_tokens < num_hash_blocks * hash_block_size` only retires hashes covering *fewer* tokens — a partial entry growing forward within its block — and is silent when the block already carries a longer one. Added a comment there saying exactly that and pointing at the caller, since as you say it reads like it already handles this case.

**Guard placement.** I kept the caller-side early return. What decided it for me: there are two callers of `cache_partial_block`, and the other one — `MambaManager._cache_partial_tail_block` — is already immune by construction. It guards with `num_tokens != latest_prompt_hash_boundary`, keyed on the *current* token count, so it fires exactly once. `FullAttentionManager` derives `boundary_tokens` from the fixed `num_prompt_tokens`, which is why it re-fires every decode step. So the hole is one caller's call pattern, not a missing primitive-level invariant, and I'd rather not hand a first PR a change to a primitive that also has to thread the legitimate-first-registration case you flagged. That said, I don't feel strongly — if a maintainer prefers it in `cache_partial_block` I'll move it; the regression test drives `FullAttentionManager.cache_blocks` so it holds either way.

**And yes, please do** — I'd appreciate you running it against your repro. Branch `fix/partial-prefix-cache-stale-hash-after-promotion` on `neetp7k9/vllm`, rebased onto `e239947777`. An independent confirmation on your original reproducer would carry more weight than my synthetic test.

**Comment by ErenAta16** [comment]
The event assertions are the right shape, and pinning "steady-state decode inside
an already-promoted block emits no events at all" is stronger than what I asked
for. That second assertion is the one that will still be meaningful in a year,
because it fails on any future change that re-enters the pool per step, not just
on a resurrection.

Putting the guard in both places is also better than either alone, and the
comment explaining why the `_cache_partial_tail_block` one is belt-and-braces
rather than redundant is worth keeping.

One thing the PR names but does not close, and I think it needs a decision rather
than a docstring. Your new docstring says callers that key follow-up bookkeeping
off the return value must treat both refusals identically, and it names
`MambaManager._cache_partial_tail_block`. That caller is not touched by the diff
and not covered by the test, which imports only `FullAttentionManager`.

Here is what it does with the value on current `main`
(`single_type_kv_cache_manager.py:1764`):

```python
partial_hash = self.block_pool.cache_partial_block(...)
if partial_hash is not None:
    self._partial_hit_reqs[request.request_id] = (block_idx, source_block)
    self.num_cached_block[request.request_id] = block_idx
    self._producer_partial_tail_reqs[request.request_id] = num_tokens
return partial_hash
```

Three pieces of bookkeeping gated on that check, and the third registers the
request as a producer so `allocate_new_blocks` hands the CoW block to the
connector for offload.

What makes this a behaviour change rather than a no-op: before your PR the only
`None` was the null-block case, and `MambaManager` already returns early on
`source_block.is_null` a few lines above, so `partial_hash is not None` was
effectively unconditional on that path. Your change gives `None` a second meaning
that this caller can actually reach, so that branch becomes conditional for the
first time, and on a promoted block all three updates are now skipped.

Skipping the partial-hit bookkeeping there may well be correct, since the
boundary really is superseded. The offload registration is the one I am less sure
about, because whether the promoted full block covers what the connector would
otherwise have taken is not something I can settle from reading. Either way it is
a decision the PR is making implicitly, in a path it does not exercise.

Not asking you to widen the scope if you would rather keep this focused. A
sentence in the PR saying which behaviour is intended for the Mamba caller would
be enough for a reviewer to agree with it deliberately.
